proposals/2020-06-26_decrypt-max-header-size-max-body-size/proposal.md

```diff
@@ -0,0 +1,275 @@
+would exhaust many memory runtime limits.
+As such, customers who are concerned with the stability and memory usage of their systems
+may care about providing stricter bounds to these fields.
+
+Note that the Encryption Context field in the header
+and the signature length field in the footer are also variable length,
+but their size only goes up to 2^16 bytes (64 KiB).
+This size is well within the bounds of memory runtime limits for the most languages,
+and thus we do not consider the bounds of these fields a concern.
+
+Note that all implementations dynamically allocate resources while processing messages.
+
+### When should users set this value?
+
+Users should set this value if they have a use case where they need to process messages of
+unknown size and those messages do not come from a trusted source.
+
+### When should users NOT set this value?
+
+This control should be considered strictly an operational control that can
+mitigate highly-bounded resource consumption for certain use cases.
+
+If a user has a security requirement to not decrypt messages with a certain content length
+or message header length
+(e.g. try to enforce a “sign only” use case by only decrypting zero length messages),
+that security requirement should be met by construction through
+use of a custom CMM as opposed to using these controls.
+
+### What value should they set it to?
+
+Users should choose reasonable values for their use case and
+the messages that they expect to decrypt.
+They should expect the memory usage of the ESDK to scale linearly with this control,
+and can performance test their application to determine
+the ESDKs memory cost in practice for their use case.
+
+### Why not just provide one control that “does the right thing” for most use cases?
+
+The two controls end up mitigating two different sorts of resource consumption concerns.
+For `Max Body Size`, this ensures that in the short term memory doesn't spike by a certain factor.
+This field can be used to ensure that the decryption of a frame or
+an unframed message doesn't immediately exhaust memory linear to a factor of that value.
+
+For `Max Header Size`, this has a fuzzier implication on memory that
+will greatly depend on the implementation,
+and represents a memory requirement that exists for the duration of the decryption operation.
+
+By separating these controls,
+we ensure that the controls are clear and simple,
+and can easily be applied to a wide variety of use cases.
+
+### Should there be a lower limit for Max Header Size?
+
+In order to allow the smallest the message header format allows
+(no AAD, one EDK but the EDK fields themselves are empty),
+any useful `Max Header Size` needs to be at least 40 bytes.
+However, future message formats might have different lower limits,
+and thus we wouldn't be able to know until we begin processing the message
+whether the value set is valid for that message or not.
+As such, there is not much value in enforcing a limit on this control.
```

Comment:
```suggestion
Because future message format revisions might have different lower limits,
we cannot know if a limit is too low until we start processing a message.
```